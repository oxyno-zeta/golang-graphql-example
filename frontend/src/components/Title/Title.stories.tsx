import React from 'react';
import { StoryFn, Meta } from '@storybook/react';
import { mdiMenu } from '@mdi/js';
import IconButton from '@mui/material/IconButton';
import SvgIcon from '@mui/material/SvgIcon';
import Title, { Props } from './Title';

export default {
  title: 'Components/Title',
  component: Title,
  args: {
    title: 'Fake',
  },
} as Meta<typeof Title>;

export const Playground: StoryFn<typeof Title> = function C(args: Props) {
  return <Title {...args} />;
};

export const LeftElement: StoryFn<typeof Title> = function C() {
  return (
    <Title
      leftElement={
        <IconButton>
          <SvgIcon>
            <path d={mdiMenu} />
          </SvgIcon>
        </IconButton>
      }
      title="Fake"
    />
  );
};

export const RightElement: StoryFn<typeof Title> = function C() {
  return (
    <Title
      rightElement={
        <IconButton>
          <SvgIcon>
            <path d={mdiMenu} />
          </SvgIcon>
        </IconButton>
      }
      title="Fake"
    />
  );
};

export const RightAndLeftElement: StoryFn<typeof Title> = function C() {
  return (
    <Title
      leftElement={
        <IconButton>
          <SvgIcon>
            <path d={mdiMenu} />
          </SvgIcon>
        </IconButton>
      }
      rightElement={
        <IconButton>
          <SvgIcon>
            <path d={mdiMenu} />
          </SvgIcon>
        </IconButton>
      }
      title="Fake"
    />
  );
};
