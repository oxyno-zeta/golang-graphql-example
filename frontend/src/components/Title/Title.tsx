import React, { ReactNode } from 'react';
import Typography, { TypographyProps } from '@mui/material/Typography';
import Toolbar, { ToolbarProps } from '@mui/material/Toolbar';

export interface Props {
  leftElement?: ReactNode;
  titleElement?: ReactNode;
  rightElement?: ReactNode;
  title?: string;
  titleTypographyProps?: Partial<TypographyProps>;
  toolbarProps?: Partial<ToolbarProps>;
}

function Title({
  leftElement = null,
  title = '',
  titleElement = null,
  rightElement = null,
  titleTypographyProps = {
    variant: 'h5',
    color: 'inherit',
    style: { fontWeight: 'bold' },
  },
  toolbarProps = {
    style: { paddingLeft: '0px', paddingRight: '0px' },
  },
}: Props) {
  return (
    <Toolbar {...toolbarProps}>
      {leftElement}
      {title && <Typography {...titleTypographyProps}>{title}</Typography>}
      {titleElement}
      <div style={{ flexGrow: 1 }} />
      {rightElement}
    </Toolbar>
  );
}

export default Title;
