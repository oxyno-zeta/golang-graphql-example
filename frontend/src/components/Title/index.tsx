import Title, { type Props as TitleProps } from './Title';

export default Title;
export type { TitleProps };
