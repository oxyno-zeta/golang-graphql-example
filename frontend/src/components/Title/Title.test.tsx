import React from 'react';
import { render } from '@testing-library/react';
// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { mdiMenu, mdiIdentifier } from '@mdi/js';
import IconButton from '@mui/material/IconButton';
import SvgIcon from '@mui/material/SvgIcon';

import Title from './Title';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

describe('Title', () => {
  it('should display title only', async () => {
    const { container } = render(<Title title="Fake title" />);
    // Now find text
    expect(container).toHaveTextContent('Fake title');
    expect(container).toMatchSnapshot();
  });

  it('should display title with left element only', async () => {
    const { container, findByRole } = render(
      <Title
        leftElement={
          <IconButton>
            <SvgIcon>
              <path d={mdiMenu} />
            </SvgIcon>
          </IconButton>
        }
        title="Fake title"
      />,
    );
    // Now find text
    expect(container).toHaveTextContent('Fake title');
    // Find button
    const buttonElement = await findByRole('button');
    expect(buttonElement).not.toBeNull();
    expect(buttonElement).toEqual(container.firstChild?.firstChild);
    expect(container).toMatchSnapshot();
  });

  it('should display title with right element only', async () => {
    const { container, findByRole } = render(
      <Title
        rightElement={
          <IconButton>
            <SvgIcon>
              <path d={mdiMenu} />
            </SvgIcon>
          </IconButton>
        }
        title="Fake title"
      />,
    );
    // Now find text
    expect(container).toHaveTextContent('Fake title');
    // Find button
    const buttonElement = await findByRole('button');
    expect(buttonElement).not.toBeNull();
    expect(buttonElement).toEqual(container.firstChild?.lastChild);
    expect(container).toMatchSnapshot();
  });

  it('should display title with right and left element', async () => {
    const { container, findAllByRole } = render(
      <Title
        leftElement={
          <IconButton>
            <SvgIcon>
              <path d={mdiMenu} />
            </SvgIcon>
          </IconButton>
        }
        rightElement={
          <IconButton>
            <SvgIcon>
              <path d={mdiIdentifier} />
            </SvgIcon>
          </IconButton>
        }
        title="Fake title"
      />,
    );
    // Now find text
    expect(container).toHaveTextContent('Fake title');
    // Find button
    const buttonElements = await findAllByRole('button');
    expect(buttonElements).not.toBeNull();
    expect(buttonElements).toHaveLength(2);
    expect(buttonElements[0]).toEqual(container.firstChild?.firstChild);
    expect(buttonElements[0].firstChild?.firstChild).toHaveAttribute('d', mdiMenu);
    expect(buttonElements[1]).toEqual(container.firstChild?.lastChild);
    expect(buttonElements[1].firstChild?.firstChild).toHaveAttribute('d', mdiIdentifier);
    expect(container).toMatchSnapshot();
  });
});
