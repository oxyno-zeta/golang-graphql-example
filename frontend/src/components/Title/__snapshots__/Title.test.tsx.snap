// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title should display title only 1`] = `
.emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
  min-height: 56px;
}

@media (min-width:600px) {
  .emotion-0 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media (min-width:0px) {
  @media (orientation: landscape) {
    .emotion-0 {
      min-height: 48px;
    }
  }
}

@media (min-width:600px) {
  .emotion-0 {
    min-height: 64px;
  }
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
  color: inherit;
}

<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular  emotion-0"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5  emotion-1"
      style="font-weight: bold;"
    >
      Fake title
    </h5>
    <div
      style="flex-grow: 1;"
    />
  </div>
</div>
`;

exports[`Title should display title with left element only 1`] = `
.emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
  min-height: 56px;
}

@media (min-width:600px) {
  .emotion-0 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media (min-width:0px) {
  @media (orientation: landscape) {
    .emotion-0 {
      min-height: 48px;
    }
  }
}

@media (min-width:600px) {
  .emotion-0 {
    min-height: 64px;
  }
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-1::-moz-focus-inner {
  border-style: none;
}

.emotion-1.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-1 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-1:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-1:hover {
    background-color: transparent;
  }
}

.emotion-1.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: 1.5rem;
}

.emotion-3 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
  color: inherit;
}

<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular  emotion-0"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <button
      class="MuiButtonBase-root  MuiIconButton-root MuiIconButton-sizeMedium  emotion-1"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium  emotion-2"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z"
        />
      </svg>
    </button>
    <h5
      class="MuiTypography-root MuiTypography-h5  emotion-3"
      style="font-weight: bold;"
    >
      Fake title
    </h5>
    <div
      style="flex-grow: 1;"
    />
  </div>
</div>
`;

exports[`Title should display title with right and left element 1`] = `
.emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
  min-height: 56px;
}

@media (min-width:600px) {
  .emotion-0 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media (min-width:0px) {
  @media (orientation: landscape) {
    .emotion-0 {
      min-height: 48px;
    }
  }
}

@media (min-width:600px) {
  .emotion-0 {
    min-height: 64px;
  }
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-1::-moz-focus-inner {
  border-style: none;
}

.emotion-1.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-1 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-1:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-1:hover {
    background-color: transparent;
  }
}

.emotion-1.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: 1.5rem;
}

.emotion-3 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
  color: inherit;
}

<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular  emotion-0"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <button
      class="MuiButtonBase-root  MuiIconButton-root MuiIconButton-sizeMedium  emotion-1"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium  emotion-2"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z"
        />
      </svg>
    </button>
    <h5
      class="MuiTypography-root MuiTypography-h5  emotion-3"
      style="font-weight: bold;"
    >
      Fake title
    </h5>
    <div
      style="flex-grow: 1;"
    />
    <button
      class="MuiButtonBase-root  MuiIconButton-root MuiIconButton-sizeMedium  emotion-1"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium  emotion-2"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 7V9H9V15H10V17H6V15H7V9H6V7H10M16 7C17.11 7 18 7.9 18 9V15C18 16.11 17.11 17 16 17H12V7M16 9H14V15H16V9Z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Title should display title with right element only 1`] = `
.emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
  min-height: 56px;
}

@media (min-width:600px) {
  .emotion-0 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media (min-width:0px) {
  @media (orientation: landscape) {
    .emotion-0 {
      min-height: 48px;
    }
  }
}

@media (min-width:600px) {
  .emotion-0 {
    min-height: 64px;
  }
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
  color: inherit;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: 1.5rem;
}

<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular  emotion-0"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5  emotion-1"
      style="font-weight: bold;"
    >
      Fake title
    </h5>
    <div
      style="flex-grow: 1;"
    />
    <button
      class="MuiButtonBase-root  MuiIconButton-root MuiIconButton-sizeMedium  emotion-2"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium  emotion-3"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
